### DISCUSSION TITLE
[@types/ramda] Issue with converge(divide, [sum, length])

### DISCUSSION THREAD

**Discussion by rpearce** [DefinitelyTyped/DefinitelyTyped#58327] [Issues with a @types package] [answered]
If you take the example from [the ramda converge docs](https://ramdajs.com/docs/#converge),

```ts
import { converge, divide, sum, length } from 'ramda'

const average = converge(divide, [sum, length])
average([1, 2, 3, 4, 5, 6, 7]) //=> 4
```

this happens, and I'm not quite sure how to get around this,

```sh
error TS2769: No overload matches this call.
  The last overload gave the following error.
    Argument of type '{ (__: Placeholder, b: number): (a: number) => number; (__: Placeholder): (b: number, a: number) => number; (a: number, b: number): number; (a: number): (b: number) => number; }' is not assignable to parameter of type '(...args: readonly [number] & { length: 1; }) => (b: number) => number'.
      Types of parameters '__' and 'args' are incompatible.
        Type 'readonly [number] & { length: 1; }' is not assignable to type '[__: Placeholder, b: number]'.

const average = converge(divide, [sum, length])
```

Versions:
```json
    "ramda": "~0.27.2"
    "@types/ramda": "~0.27.64",
```

* * *

On further investigation, this _seems_ to work, but it seems a bit heavy handed...

```ts
const average = converge(
  (...args: readonly number[]) => divide(args[0], args[1]),
  [sum, length]
)
average([1, 2, 3, 4, 5, 6, 7]) //=> 4
```

[a comment by an automated account is not shown]

**Comment by adispring** [comment]
Maybe we should make `converge` and other apis more loose?

**Reply by rpearce** [reply]
Perhaps? Another thing that I found was that `propOr` required you to pass two arguments at the start, but ramda's auto-currying allows you to pass them as unary function calls, like `propOr(0)('length')([1, 2, 3])`, which is useful for partially applying arguments, storing the result, and using getting reuse out of that.

```js
const propOr0    = propOr(0)
const safeLength = propOr0('length')

safeLength([1, 2, 3])
```

Would this be an example of "make ... other apis more loose"?

**Comment by adispring** [comment]
If Typescript support customizing compile Type Error, this will be easy to optimize. But until now, typescript has not support custom type error.

Or we can do some trick to throw compile error, as follows: 

https://github.com/microsoft/TypeScript/issues/23689
https://github.com/microsoft/TypeScript/pull/40468

@Nicholaiii @NEWESTERS

**Comment by valerii15298** [comment, marked as answer]
Hello @rpearce,
now you can do it like this:
```ts
import { converge, divide, sum, length } from 'ramda';

const average = converge((...args) => divide(...args), [sum, length] as const);
average([1, 2, 3, 4, 5, 6, 7]) //=> 4
```
Also if you wanna more examples, [here](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/ramda/test/converge-tests.ts) is some.
